feat: allow overriding the session ID

jcmfernandes opened the pull request:
<!-- Type of change
Please label this PR with one of the existing labels, depending on the scope of your change
-->

## What does this PR do?
Makes the session ID configurable, allowing for long-running sessions.

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, etc.
-->

## Why is it important?
Speaking of my use-case, Apache Pulsar takes a long time to boot, so I wanted testcontainers to boot it up, eventually take it down, but allow me to use the same container across several runs. This wasn't the case, since I was getting a new reaper running on every test run. I wanted a rolling-window behavior, hence this change.

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- Closes #2804

<!-- Recommended
## How to test this PR

Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Follow-ups

Update the documentation.

jcmfernandes commented:
@stevenh yes, because I want to keep the testcontainer running for the entirety of my work session.

I usually start a watcher that runs tests automatically as I save changes. I want the testcontainer to boot on the first test run and to stay around for, say, one hour, waiting for new connections. Furthermore, I then want new connections to reset that timeout. Example.

1. Run a test that starts the testcontainer at 9:05 AM with a reconnection timeout of 1 hour. Ryuk must wait for new connections until 10:05 AM, and GC the testcontainer if that doesn't happen.
2. Run another test at 9:35 AM --> I now want ryuk to wait until 10:35 AM for new connections.

This is precisely the behavior I get by manually setting the session ID. Now, I'm happy to discuss ways to make the session ID stable and auto-generated.

stevenh commented:
Thanks again for the clarification, the challenge I see with that is not only does it impact the long lived container but all containers for example.

You run a test which starts the long lived DB container as well a short lived test container. The desired behaviour is the DB running container stays running but the test container exits after the test is run. However in the case where the test fails and the short lived contain didn't exit as expected, with a standard ryuk setup this wouldn't be an issue as the test container would be quickly cleaned up, but with `RYUK_RECONNECTION_TIMEOUT` set to one hour no clean up will happen until that hour expires.

Does this make sense?

To handle this situation properly I think we'd need more fine grained control of when specific containers are cleaned up.

jcmfernandes commented:
Thanks for the prompt reply, @stevenh.

That's a fair concern — it seems that testcontainers is missing some kind of "group" concept to which certain GC options apply. Currently, that's a "session", and this PR won't change it. That is, I don't see how this PR makes things worse regarding your concern, and it solves an actual reported issue.

mdelapenya reviewed (commented):
Hi @jcmfernandes, thanks for opening this PR, and following-up after @kiview's comments. I'm returning today after summer break, working on the open PRs/issues in the repo.

I'm not against the changes in this PR, although it introduces some breaking changes (BCs) after removing public API functions. See my comments for that. We could avoid it offering a deprecation path when/if possible, although we sometimes introduce BCs if needed.

I'd rename the PR to `feat: allow overriding the sessionID` and would focus on that, only. The added value if we do this is we would also enable Bazel folks to contribute their own env var for the session ID at build time. Nevertheless, we must check how the reaper behaves in that scenario. As discussed here, it's by design that the reaper is spawned just once per test session (a shell), but it's true that the design could be inconsistent with reusable containers used in multiple shells (different session IDs). Sharing state, even in tests, is a complex thing to maintain, so we probably need a more robust design for both features: reusable containers and the reaper, and it could need dedicated work on that direction.

@kiview @stevenh wdyt?

mdelapenya commented on internal/core/bootstrap.go:
```diff
@@ -8,6 +8,7 @@ import (
 
 	"github.com/google/uuid"
 	"github.com/shirou/gopsutil/v4/process"
+	"github.com/testcontainers/testcontainers-go/internal/config"
```
bug: linter is complaining about this line. Could you run `make lint` to address that? 🙏

stevenh commented on internal/core/bootstrap.go:
```diff
@@ -45,6 +47,11 @@ var processID string
 const sessionIDPlaceholder = "testcontainers-go:%d:%d"
 
 func init() {
+	cfg := config.Read()
```
@mdelapenya could calling `config.Read()` in `init` cause undesired behaviour due to how early it called?

I'm thinking this would prevent things like changing the environment variables in a test from having the desired effect as an example.

jcmfernandes replied:
Thanks for the feedback, @stevenh. I understand your concern: reading the config in an `init()` function felt wrong.

I change the code so that `config.Read` falls back to the session in `core/bootstrap` (I moved `bootstrap.go` to a new package to avoid import cycle between `internal/config` and ``internal/core`). Please let me know what you think.

stevenh commented on internal/core/bootstrap.go:
```diff
@@ -57,7 +64,9 @@ func init() {
 	processes, err := process.Processes()
 	if err != nil {
-		sessionID = uuid.New().String()
+		if sessionID == "" {
+			sessionID = uuid.New().String()
+		}
```
suggestion: put this in a `defer` check in an else block to the `cfg.SessionID != ""` less to maintain that way.

jcmfernandes replied:
I'm no longer pushing this change.

stevenh reviewed (changes requested):
I'm concerned calling `config.Read` in `init` could have some unintended side effects which are impossible to address, something we'll need to confirm.

mdelapenya reviewed (dismissed):
LGTM! I'm approving it even though I added a comment that would be good to add.

Once addressed, I think we can re-run the CI and merge.

Thanks!

PS: Many times, we are forced to do multitasking, so please sorry for taking too long to address this.

jcmfernandes commented on internal/config/config.go:
```diff
@@ -121,6 +130,13 @@ func read() Config {
+			config.SessionID = sessionID
+		} else if config.SessionID == "" {
+			config.SessionID = bootstrap.SessionID()
+		}
```
The "else if" comparison is indeed irrelevant. Fixing that. Thank you!

mdelapenya commented:
Some tests are failing on CI, can you check?

jcmfernandes commented:
@mdelapenya mind giving this a final look? Thank you so much.

mdelapenya commented:
@jcmfernandes can you rebase from current main 🙏 Fine if you push-force, as the changeset's size is reduced

BTW I'm cutting a new release soon, and this work will land into it (sorry for the late review 😞 )

jcmfernandes committed:
chore: lint

jcmfernandes committed:
refactor: have the config load from core/bootstrap

jcmfernandes committed:
chore: lint

jcmfernandes committed:
refactor: tidy

jcmfernandes committed:
test: fix description

jcmfernandes committed:
feat: allow overriding the session ID

jcmfernandes commented:
Rebased @mdelapenya, thank you! :raised_hands: 

I opened #3837 as it was raised the [LLM code review](https://github.com/testcontainers/testcontainers-go/pull/3051#pullrequestreview-4856522740) added to this branch and it's a genuine bug.

mdelapenya commented:
@jcmfernandes for some reason I cannot update your PR branches. Could you update them yourself? I already fixed the lint error for the azurite module

jcmfernandes commented:
> @jcmfernandes for some reason I cannot update your PR branches. Could you update them yourself? I already fixed the lint error for the azurite module

Rebased on main.

mdelapenya commented:
CI fails, could you take a look?

jcmfernandes commented:
> CI fails, could you take a look?

Done. It was something wonky that stem from the review but truly is necessary.

mdelapenya commented on internal/config/config.go:
```diff
@@ -121,6 +130,13 @@ func read() Config {
+		sessionID := os.Getenv("TESTCONTAINERS_SESSION_ID")
+		if sessionID != "" {
+			config.SessionID = sessionID
+		} else if config.SessionID == "" {
```
question: if this function is called just once , who could have set the session ID value here? Could this be a sympton the tests are kind of setting a state and this elseIf path is exactly for that?

jcmfernandes replied:
Clarifying: are you refering to https://github.com/testcontainers/testcontainers-go/actions/runs/31088981575/job/92592530979?pr=3051#step:8:5482 ?

If so, env vars take priority over the values included in the properties file - that is, we **first** load `config` with values from the properties file and only then we **override** them with values coming from env vars. Therefore, if we remove the `config.SessionID == ""` check, then we **always** override the value coming from the properties file when the env var isn't set.

Does this make sense?

mdelapenya replied:
Oh I see, it's the properties the one that could have set it, sorry, Now I get it, thanks for the clarification

mdelapenya reviewed (approved):
LGTM, thanks
